src/file_pipe_log/pipe_builder.rs

```diff
@@ -49,9 +175,18 @@ impl<B: FileBuilder> DualPipesBuilder<B> {
             listeners,
             append_files: Vec::new(),
             rewrite_files: Vec::new(),
+            extra_params: None,
         }
     }
 
+    pub fn set_truncate_params(&mut self, truncate_params: Option<TruncateQueueParameter>) {
```

Comment:
I'm a bit confused why is it so difficult to hide params from the builder. You can determine which queue to replay before entering `recover_queue`.

Reply:
> You can determine which queue to replay before entering recover_queue 

Yes, it can be determined before entering `recover_queue `, as `recover` return `Result<ReplayMachine>`, maybe we should write as follower

```
if should_recover_queue() {
}
else {
   Ok(machine)
}
```
I think it's not very nice to add the else block with simple return Ok,  so i put the logic into `recover_queue`


> I'm a bit confused why is it so difficult to hide params from the builder 

I can't get your idea clearly. you mean we should add `truncate_paramter` into default constructor of `Builder` or it's not wise to place the `truncate_paramter` in `DualPipesBuilder` ?

Reply:
Builder shouldn't know about any parameters specific to the machine type. You should figure out the log queue and call corresponding `recover_queue` in `Engine::truncate`.